### PR TITLE
Christian sf 105

### PR DESCRIPTION
**Pull request by christianjosephvarriale**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -113,7 +113,10 @@ Action creates a single object. Input metadata is fetched dynamically from your
 * **Allow criteria to be omitted** - checkbox, if checked - search criteria can be omitted and the empty object will be returned, else - search criteria are required.
 * **Allow zero results** - checkbox, if checked and nothing is found - empty object will be returned, else - action throw an error.
 * **Pass binary data to the next component (if found object has it)** - a checkbox, if it is checked and found object has a binary field (type of base64) then its data will be passed to the next component as a binary attachment.
+<<<<<<< HEAD
+=======
```
Comment (jhorbulyk) [inline README.md]: Fix merge conflict.

File: .vscode/launch.json [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+            "type": "node",
+            "request": "launch",
+            "name": "Launch Program",
+            "skipFiles": [
+                "<node_internals>/**"
+            ],
+            "program": "${workspaceFolder}/index.js"
+        }
+    ]
+}
```
Comment (jhorbulyk) [inline .vscode/launch.json]: Most of the team does not use VS Code.  Oh the one hand, this file conatins useful info.  On the other hand, it's unlikely to be used.

File: README.md [hunk 3]
```diff
@@ -122,12 +125,15 @@ Metadata contains one field whose name, type and mandatoriness are generated acc
 #### Limitations
 When **Pass binary data to the next component (if found object has it)** is checked and this action is used with Local Agent error would be thrown: 'getaddrinfo ENOTFOUND steward-service.platform.svc.cluster.local steward-service.platform.svc.cluster.local:8200'
 
+<<<<<<< HEAD
+=======
```
Comment (jhorbulyk) [inline README.md]: More merge conflicts

File: README.md [hunk 4]
```diff
@@ -136,13 +142,16 @@ Lookup a list of objects satisfying specified criteria.
...
 * **Enable Cache Usage** - Flag to enable cache usage.
 
 #### Note
 Action has caching mechanism. By default action stores last 10 request-response pairs for 10 min duration.
 This parameters can be changed by setting environment variables:
 * **HASH_LIMIT_TIME** - Hash expiration time in milis
 * **HASH_LIMIT_ELEMENTS** - Hash size number limit
+>>>>>>> 865ca3a2219cbc12f479ab30f9a1ac8ea340852f
```
Comment (jhorbulyk) [inline README.md]: Can you add remarks about your change to the readme?

File: component.json [hunk 5]
```diff
@@ -255,20 +255,37 @@
       }
     },
     "delete": {
-      "title": "Delete Object",
+      "title": "Delete Object (at most 1)",
       "main": "./lib/actions/deleteObject.js",
       "description": "Delete Selected Object",
+      "dynamicMetadata": true,
```
Comment (jhorbulyk) [inline component.json]: If you use dynamic metadata, then you should remove the `metadata` section for the action (as both parts will now come from the dynamic metadata.

File: lib/actions/deleteObject.js [hunk 6]
```diff
@@ -1,31 +1,101 @@
 const { messages } = require('elasticio-node');
 const MetaLoader = require('../helpers/metaLoader');
 const sfConnection = require('../helpers/sfConnection.js');
+const helpers  = require('../helpers/deleteObjectHelpers.js')
+// "test": "eslint & mocha spec --recursive --timeout 50000",
+// "integration-test": "mocha spec-integration --recursive --timeout 50000"
```
Comment (jhorbulyk) [inline lib/actions/deleteObject.js]: Remove

File: lib/actions/deleteObject.js [hunk 7]
```diff
@@ -1,31 +1,101 @@
...
+ * @param configuration encapsulates the lookup value and lookup key
+*/
+module.exports.process = async function process(message, configuration) {
+  const { lookupField } = configuration;
+  const lookupValue = message.body[lookupField];
+  const res = [];
+  const sfConn = sfConnection.createConnection(configuration, this);
+
+  this.logger.info(`Heres the ouath ${configuration.oauth.refresh_token} refresh token`);
+  this.logger.info(`Heres the ouath ${configuration.oauth.access_token} refresh token`);
```
Comment (jhorbulyk) [inline lib/actions/deleteObject.js]: Don't log at info level.

File: spec-integration/actions/secrets.env [hunk 8]
```diff
@@ -0,0 +1,4 @@
+ACCESS_TOKEN=00DE0000000dwKc!ARcAQCVuf0C.Qk7f1gHDjns.PSu3vPSbCofmaGx7xVgdCSXAMIQS00_z8MlKMSnlTg0DmRd5Lz0oM0nONwY0Sij2GsZPTlRM
```
Comment (jhorbulyk) [inline spec-integration/actions/secrets.env]: Don't check this in.

File: lib/actions/deleteObject.js [hunk 9]
```diff
@@ -1,31 +1,101 @@
...
-  } catch (err) {
-    this.logger.error(`Salesforce error. ${err.message}`);
-    return messages.newMessageWithBody({ });
-  }
+  const meta = await sfConn.describe(configuration.sobject);
+  const field = meta.fields.find(fld => fld.name === lookupField);
+  this.logger.info(MetaLoader.TYPES_MAP[field.type]);
+  const condition = (['date','datetime'].includes(field.type) || MetaLoader.TYPES_MAP[field.type] === 'number')
+    ? `${lookupField} = ${lookupValue}`
+    : `${lookupField} = '${lookupValue}'`;
```
Comment (jhorbulyk) [inline lib/actions/deleteObject.js]: Is there re-usability from the lookup object case?

File: lib/actions/deleteObject.js [hunk 10]
```diff
@@ -1,31 +1,101 @@
...
+  await sfConn.sobject(configuration.sobject)
+    .select('*')
+    .where(condition)
+    .on('record', (record) => {
+      res.push(record);
+    })
+    .on('end', async () => {
+      if (res.length === 0) {
+
+        const err = new Error('No objects found');
```
Comment (jhorbulyk) [inline lib/actions/deleteObject.js]: When nothing matching is found, then I think the desired behavior is to emit the empty message, not throw an error.

File: spec-integration/actions/deleteObject.spec.js [hunk 11]
```diff
@@ -0,0 +1,96 @@
...
+    emit: sinon.spy(),
+    logger,
+  };
+
+  it('Correctly throws on a non-existent Contact', async () => {
+    const config = deleteObjectData.cases[0].config;
+    config.oauth.refresh_token = process.env.REFRESH_TOKEN;
+    config.oauth.access_token = process.env.ACCESS_TOKEN;
+    await deleteObject.process.call(emitter, deleteObjectData.cases[0].message, config)
+      .then(() => {
```
Comment (jhorbulyk) [inline spec-integration/actions/deleteObject.spec.js]: Don't use then with await.

File: spec-integration/actions/deleteObject.json [hunk 12]
```diff
@@ -0,0 +1,128 @@
+{
```
Comment (jhorbulyk) [inline spec-integration/actions/deleteObject.json]: Does something populate the SF instance with this data before the integration tests delete it?

File: .vscode/launch.json (hunk 2 above)
Comment (jhorbulyk) [reply .vscode/launch.json]: Can you remove this file?

File: lib/actions/deleteObject.js (hunk 9 above)
Comment (jhorbulyk) [reply lib/actions/deleteObject.js]: @christianjosephvarriale Ping?

File: spec-integration/actions/deleteObject.spec.js (hunk 11 above)
Comment (jhorbulyk) [reply spec-integration/actions/deleteObject.spec.js]: @christianjosephvarriale Ping

File: lib/actions/deleteObject.js (hunk 9 above)
Comment (christianjosephvarriale) [reply lib/actions/deleteObject.js]: @jhorbulyk yes this is the code from the lookup case

File: spec-integration/actions/deleteObject.spec.js (hunk 11 above)
Comment (christianjosephvarriale) [reply spec-integration/actions/deleteObject.spec.js]: it's been removed

File: .vscode/launch.json (hunk 2 above)
Comment (christianjosephvarriale) [reply .vscode/launch.json]: removed

File: spec-integration/actions/secrets.env (hunk 8 above)
Comment (christianjosephvarriale) [reply spec-integration/actions/secrets.env]: removed

**Pull request closed without merging**